### PR TITLE
Feat/calldata pi proof

### PR DESCRIPTION
**Pull request by ThomasPiellard**
plonk verifier
-------------

* added @julien-marchand modifications to store proof and public inputs in calldata instead of memory
* used calldatacopy when needed instead of mstore(mload(..))

### CODE REVIEW

File: backend/plonk/bn254/solidity.go [hunk 1]
```diff
@@ -708,22 +700,25 @@ contract PlonkVerifier {
...
         fr_acc_mul(folded_evals, add(aproof, proof_grand_product_at_zeta_omega), random)
 
         let folded_evals_commit := mPtr
         mPtr := add(folded_evals_commit, 0x40)
-        mstore(folded_evals_commit, {{ fpstr .Kzg.G1.X }})
-        mstore(add(folded_evals_commit, 0x20), {{ fpstr .Kzg.G1.Y }})
+        mstore(folded_evals_commit, 14312776538779914388377568895031746459131577658076416373430523308756343304251)
+        mstore(
+          add(folded_evals_commit, 0x20),
+          11763105256161367503191792604679297387056316997144156930871823008787082098465
```
Comment (ivokub) [inline backend/plonk/bn254/solidity.go:718]: use template.

File: backend/plonk/bn254/solidity.go [hunk 2]
```diff
@@ -1071,7 +1094,18 @@ contract PlonkVerifier {
...
+        let state := mload(0x40)
+        mstore(mPtr, mload(p))
+        mstore(add(mPtr, 0x20), mload(add(p, 0x20)))
+        mstore(add(mPtr, 0x40), calldataload(q))
+        mstore(add(mPtr, 0x60), calldataload(add(q, 0x20)))
+        let l_success := staticcall(gas(), 6, mPtr, 0x80, dst, 0x40)
+        mstore(add(state, state_success), and(l_success, mload(add(state, state_success))))
+      }
+
+      // // dst <- [s]src
```
Comment (ivokub) [inline backend/plonk/bn254/solidity.go:1108]: double comment.

File: backend/plonk/bn254/solidity.go [hunk 3]
```diff
@@ -1071,7 +1094,18 @@ contract PlonkVerifier {
         mstore(add(state, state_success), and(l_success,mload(add(state, state_success))))
       }
 
-      // dst <- [s]src
+      function point_add_calldata(dst, p, q, mPtr) {
+        // let mPtr := add(mload(0x40), state_last_mem)
```
Comment (ivokub) [inline backend/plonk/bn254/solidity.go:1098]: deadline

File: backend/plonk/bn254/solidity.go [hunk 4]
```diff
@@ -1082,6 +1116,17 @@ contract PlonkVerifier {
         mstore(add(state, state_success), and(l_success,mload(add(state, state_success))))
       }
 
+      // dst <- [s]src
+      function point_mul_calldata(dst, src, s, mPtr) {
+        // let mPtr := add(mload(0x40), state_last_mem)
```
Comment (ivokub) [inline backend/plonk/bn254/solidity.go:1121]: deadline

File: backend/plonk/bn254/solidity.go [hunk 5]
```diff
@@ -1116,6 +1174,7 @@ contract PlonkVerifier {
         res := mload(mPtr)
       }
     }
+    success = true;
```
Comment (ivokub) [inline backend/plonk/bn254/solidity.go:1177]: Why this?

File: backend/plonk/bn254/solidity.go [hunk 6]
```diff
@@ -251,49 +230,82 @@ contract PlonkVerifier {
...
+          input_checks := and(input_checks,lt(calldataload(p), r_mod))
+          p := add(p, 0x20)
+        }
+        if iszero(input_checks) {
+          error_inputs_size()
+        }
+      }
+
+      // the 'a' prepending proof in aproof stands for 'assembly'. It's just to not use again the name proof (not allowed)
+      function check_proof_size(actual_proof_size, aproof) {
```
[the remark this answers is not in the document]
Comment (ThomasPiellard) [reply backend/plonk/bn254/solidity.go:260]: ah shoot nice catch

File: backend/plonk/bn254/solidity.go (hunk 5 above)
Comment (ThomasPiellard) [reply backend/plonk/bn254/solidity.go]: an enormous mistake that made me think it was working while there was a bug^^ I fixed it last friday

**Review by ivokub — APPROVED** [review]
Looks good now.

**Pull request merged**